From: Fumitoshi UKAI <ukai@debian.or.jp>
Date: Tue, 19 Oct 2004 02:27:15 +0900
Subject: grep -i breaks character classes for utf-8 locales

fix the following problem in multibyte locales.
 % echo Y | egrep -i '[y]'
 %

derived from gawk's dfa.c.

Reported-by: Heike C. Zimmerer <hcz@hczim.de>
Bug-Debian: http://bugs.debian.org/249245
---
 src/dfa.c |   14 ++++++++++++++
 1 files changed, 14 insertions(+), 0 deletions(-)

diff --git a/src/dfa.c b/src/dfa.c
index 3184dd1..da0206b 100644
--- a/src/dfa.c
+++ b/src/dfa.c
@@ -700,6 +700,20 @@ parse_bracket_exp_mb ()
 	  REALLOC_IF_NECESSARY(work_mbc->chars, wchar_t, chars_al,
 			       work_mbc->nchars + 1);
 	  work_mbc->chars[work_mbc->nchars++] = (wchar_t)wc;
+	  if (case_fold && (iswlower((wint_t) wc) || iswupper((wint_t) wc)))
+	    {
+		wint_t altcase;
+
+		altcase = wc;		/* keeps compiler happy */
+		if (iswlower((wint_t) wc))
+		  altcase = towupper((wint_t) wc);
+		else if (iswupper((wint_t) wc))
+		  altcase = towlower((wint_t) wc);
+
+		REALLOC_IF_NECESSARY(work_mbc->chars, wchar_t, chars_al,
+			       work_mbc->nchars + 1);
+		work_mbc->chars[work_mbc->nchars++] = (wchar_t) altcase;
+	    }
 	}
     }
   while ((wc = wc1) != L']');
-- 
1.6.6

